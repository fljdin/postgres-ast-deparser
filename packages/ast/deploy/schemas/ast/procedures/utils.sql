-- Deploy schemas/ast/procedures/utils to pg

-- requires: schemas/ast/schema

BEGIN;


COMMIT;
