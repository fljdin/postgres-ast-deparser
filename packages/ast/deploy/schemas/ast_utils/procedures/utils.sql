-- Deploy schemas/ast_utils/procedures/utils to pg

-- requires: schemas/ast_utils/schema

BEGIN;

CREATE FUNCTION ast_utils.interval (
  n int
) returns text[] as $$
	select (CASE 
WHEN ( n = 2 ) THEN ARRAY[ 'month' ]
WHEN ( n = 4 ) THEN ARRAY[ 'year' ]
WHEN ( n = 6 ) THEN ARRAY[ 'year', 'month' ]
WHEN ( n = 8 ) THEN ARRAY[ 'day' ]
WHEN ( n = 1024 ) THEN ARRAY[ 'hour' ]
WHEN ( n = 1032 ) THEN ARRAY[ 'day', 'hour' ]
WHEN ( n = 2048 ) THEN ARRAY[ 'minute' ]
WHEN ( n = 3072 ) THEN ARRAY[ 'hour', 'minute' ]
WHEN ( n = 3080 ) THEN ARRAY[ 'day', 'minute' ]
WHEN ( n = 4096 ) THEN ARRAY[ 'second' ]
WHEN ( n = 6144 ) THEN ARRAY[ 'minute', 'second' ]
WHEN ( n = 7168 ) THEN ARRAY[ 'hour', 'second' ]
WHEN ( n = 7176 ) THEN ARRAY[ 'day', 'second' ]
WHEN ( n = 32767 ) THEN ARRAY[]::text[]
END);
$$  
LANGUAGE 'sql' IMMUTABLE;

CREATE FUNCTION ast_utils.reserved (
  str text
) returns boolean as $$
	select exists( select 1 from pg_get_keywords() where catcode = 'R' AND word=str  );
$$  
LANGUAGE 'sql' SECURITY DEFINER;

CREATE FUNCTION ast_utils.objtypes ()
returns text[] as $$
	select ARRAY[ 'ACCESS METHOD', 'AGGREGATE', NULL, NULL, NULL, 'CAST', 'COLUMN', 'COLLATION', 'CONVERSION', 'DATABASE', NULL, NULL, 'DOMAIN', 'CONSTRAINT', NULL, 'EXTENSION', 'FOREIGN DATA WRAPPER', 'SERVER', 'FOREIGN TABLE', 'FUNCTION', 'INDEX', 'LANGUAGE', 'LARGE OBJECT', 'MATERIALIZED VIEW', 'OPERATOR CLASS', 'OPERATOR', 'OPERATOR FAMILY', 'POLICY', NULL, NULL, 'ROLE', 'RULE', 'SCHEMA', 'SEQUENCE', NULL, 'STATISTICS', 'CONSTRAINT', 'TABLE', 'TABLESPACE', 'TRANSFORM', 'TRIGGER', 'TEXT SEARCH CONFIGURATION', 'TEXT SEARCH DICTIONARY', 'TEXT SEARCH PARSER', 'TEXT SEARCH TEMPLATE', 'TYPE', NULL, 'VIEW' ]::text[];
$$  
LANGUAGE 'sql' IMMUTABLE;

CREATE FUNCTION ast_utils.constrainttype_idxs (typ text)
returns int as $$
  select (CASE
WHEN (typ = 'CONSTR_NULL') THEN 0
WHEN (typ = 'CONSTR_NOTNULL') THEN 1
WHEN (typ = 'CONSTR_DEFAULT') THEN 2
WHEN (typ = 'CONSTR_IDENTITY') THEN 3
WHEN (typ = 'CONSTR_CHECK') THEN 4
WHEN (typ = 'CONSTR_PRIMARY') THEN 5
WHEN (typ = 'CONSTR_UNIQUE') THEN 6
WHEN (typ = 'CONSTR_EXCLUSION') THEN 7
WHEN (typ = 'CONSTR_FOREIGN') THEN 8
WHEN (typ = 'CONSTR_ATTR_DEFERRABLE') THEN 9
WHEN (typ = 'CONSTR_ATTR_NOT_DEFERRABLE') THEN 10
WHEN (typ = 'CONSTR_ATTR_DEFERRED') THEN 11
WHEN (typ = 'CONSTR_ATTR_IMMEDIATE') THEN 12
END);
$$
LANGUAGE 'sql' IMMUTABLE;

CREATE FUNCTION ast_utils.constrainttypes (contype int)
returns text as $$
  select (CASE
WHEN (contype =  0 ) THEN 'NULL'
WHEN (contype =  1 ) THEN 'NOT NULL'
WHEN (contype =  2 ) THEN 'DEFAULT'
WHEN (contype =  4 ) THEN 'CHECK'
WHEN (contype =  5 ) THEN 'PRIMARY KEY'
WHEN (contype =  6 ) THEN 'UNIQUE'
WHEN (contype =  7 ) THEN 'EXCLUDE'
WHEN (contype =  8 ) THEN 'REFERENCES'
END);
$$
LANGUAGE 'sql' IMMUTABLE;

CREATE FUNCTION ast_utils.objtypes_idxs (typ text)
returns int as $$
	select (CASE
WHEN (typ = 'OBJECT_ACCESS_METHOD') THEN 0
WHEN (typ = 'OBJECT_AGGREGATE') THEN 1
WHEN (typ = 'OBJECT_AMOP') THEN 2
WHEN (typ = 'OBJECT_AMPROC') THEN 3
WHEN (typ = 'OBJECT_ATTRIBUTE') THEN 4
WHEN (typ = 'OBJECT_CAST') THEN 5
WHEN (typ = 'OBJECT_COLUMN') THEN 6
WHEN (typ = 'OBJECT_COLLATION') THEN 7
WHEN (typ = 'OBJECT_CONVERSION') THEN 8
WHEN (typ = 'OBJECT_DATABASE') THEN 9
WHEN (typ = 'OBJECT_DEFAULT') THEN 10
WHEN (typ = 'OBJECT_DEFACL') THEN 11
WHEN (typ = 'OBJECT_DOMAIN') THEN 12
WHEN (typ = 'OBJECT_DOMCONSTRAINT') THEN 13
WHEN (typ = 'OBJECT_EVENT_TRIGGER') THEN 14
WHEN (typ = 'OBJECT_EXTENSION') THEN 15
WHEN (typ = 'OBJECT_FDW') THEN 16
WHEN (typ = 'OBJECT_FOREIGN_SERVER') THEN 17
WHEN (typ = 'OBJECT_FOREIGN_TABLE') THEN 18
WHEN (typ = 'OBJECT_FUNCTION') THEN 19
WHEN (typ = 'OBJECT_INDEX') THEN 20
WHEN (typ = 'OBJECT_LANGUAGE') THEN 21
WHEN (typ = 'OBJECT_LARGEOBJECT') THEN 22
WHEN (typ = 'OBJECT_MATVIEW') THEN 23
WHEN (typ = 'OBJECT_OPCLASS') THEN 24
WHEN (typ = 'OBJECT_OPERATOR') THEN 25
WHEN (typ = 'OBJECT_OPFAMILY') THEN 26
WHEN (typ = 'OBJECT_POLICY') THEN 27
WHEN (typ = 'OBJECT_PUBLICATION') THEN 28
WHEN (typ = 'OBJECT_PUBLICATION_REL') THEN 29
WHEN (typ = 'OBJECT_ROLE') THEN 30
WHEN (typ = 'OBJECT_RULE') THEN 31
WHEN (typ = 'OBJECT_SCHEMA') THEN 32
WHEN (typ = 'OBJECT_SEQUENCE') THEN 33
WHEN (typ = 'OBJECT_SUBSCRIPTION') THEN 34
WHEN (typ = 'OBJECT_STATISTIC_EXT') THEN 35
WHEN (typ = 'OBJECT_TABCONSTRAINT') THEN 36
WHEN (typ = 'OBJECT_TABLE') THEN 37
WHEN (typ = 'OBJECT_TABLESPACE') THEN 38
WHEN (typ = 'OBJECT_TRANSFORM') THEN 39
WHEN (typ = 'OBJECT_TRIGGER') THEN 40
WHEN (typ = 'OBJECT_TSCONFIGURATION') THEN 41
WHEN (typ = 'OBJECT_TSDICTIONARY') THEN 42
WHEN (typ = 'OBJECT_TSPARSER') THEN 43
WHEN (typ = 'OBJECT_TSTEMPLATE') THEN 44
WHEN (typ = 'OBJECT_TYPE') THEN 45
WHEN (typ = 'OBJECT_USER_MAPPING') THEN 46
WHEN (typ = 'OBJECT_VIEW') THEN 47
END);
$$  
LANGUAGE 'sql' IMMUTABLE;

CREATE FUNCTION ast_utils.getgrantobject (node jsonb)
returns text as $$
DECLARE
  objtype int;
  targtype int;
BEGIN 

  objtype = (node->'objtype')::int;
  IF (node->'targtype') IS NOT NULL THEN
    targtype = (node->'targtype')::int;
  END IF;

  IF (objtype = 0) THEN
    RETURN 'COLUMN';
  ELSIF (objtype = 1) THEN
    IF (targtype = 1) THEN 
      RETURN 'ALL TABLES IN SCHEMA';
    ELSIF (targtype = 2) THEN 
      RETURN 'TABLES';
    END IF;
    -- TODO could be a view
    RETURN 'TABLE';
  ELSIF (objtype = 2) THEN
    RETURN 'SEQUENCE';
  ELSIF (objtype = 3) THEN
    RETURN 'DATABASE';
  ELSIF (objtype = 4) THEN
    RETURN 'DOMAIN';
  ELSIF (objtype = 5) THEN
    RETURN 'FOREIGN DATA WRAPPER';
  ELSIF (objtype = 6) THEN
    RETURN 'FOREIGN SERVER';
  ELSIF (objtype = 7) THEN
    IF (targtype = 1) THEN 
      RETURN 'ALL FUNCTIONS IN SCHEMA';
    ELSIF (targtype = 2) THEN 
      RETURN 'FUNCTIONS';
    END IF;
    RETURN 'FUNCTION';
  ELSIF (objtype = 8) THEN
    RETURN 'LANGUAGE';
  ELSIF (objtype = 9) THEN
    RETURN 'LARGE OBJECT';
  ELSIF (objtype = 10) THEN
    RETURN 'SCHEMA';
  ELSIF (objtype = 11) THEN
    RETURN 'TABLESPACE';
  ELSIF (objtype = 12) THEN
    RETURN 'TYPE';
  END IF;

  RAISE EXCEPTION 'BAD_EXPRESSION %', 'GrantObjectType';

END;
$$  
LANGUAGE 'plpgsql' IMMUTABLE;

COMMIT;
