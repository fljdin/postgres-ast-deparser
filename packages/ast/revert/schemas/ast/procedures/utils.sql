-- Revert schemas/ast/procedures/utils from pg

BEGIN;

DROP FUNCTION ast.utils;

COMMIT;
