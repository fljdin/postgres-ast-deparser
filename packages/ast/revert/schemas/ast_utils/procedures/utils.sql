-- Revert schemas/ast_utils/procedures/utils from pg

BEGIN;

DROP FUNCTION ast_utils.utils;

COMMIT;
