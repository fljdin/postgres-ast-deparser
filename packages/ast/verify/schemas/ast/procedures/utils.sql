-- Verify schemas/ast/procedures/utils  on pg

BEGIN;

SELECT verify_function ('ast.utils');

ROLLBACK;
