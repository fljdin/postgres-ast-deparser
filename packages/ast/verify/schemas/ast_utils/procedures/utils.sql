-- Verify schemas/ast_utils/procedures/utils  on pg

BEGIN;

SELECT verify_function ('ast_utils.utils');

ROLLBACK;
