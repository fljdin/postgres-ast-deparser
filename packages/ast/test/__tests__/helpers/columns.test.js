import { getConnections } from '../../utils';
import cases from 'jest-in-case';

let db, teardown;
const objs = {
  tables: {}
};

beforeAll(async () => {
  ({ db, teardown } = await getConnections());
  await db.begin();
  await db.savepoint();
});
afterAll(async () => {
  try {
    //try catch here allows us to see the sql parsing issues!
    await db.rollback();
    await db.commit();
    await teardown();
  } catch (e) {
    console.log(e);
  }
});

it('geometry', async () => {
  const [result] = await db.any(
    `
        select deparser.deparse(
          ast_helpers.column_geotype_identifier(
            v_column_geo_type := $1::text,
            v_column_geo_num := $2::int
          )
        );
          `,
    ['polygon', 4326]
  );
  expect(result).toMatchSnapshot();
});

it('arrays', async () => {
  const [result] = await db.any(
    `
        select deparser.deparse(
          ast_helpers.alter_table_add_column(
            v_schema_name := 'v_schema_name',
            v_table_name := 'v_table_name',
            v_column_name := 'v_column_name',
            v_column_type := $1::text,
            v_is_array := $2::bool
          )
        );
          `,
    ['uuid', true]
  );
  expect(result).toMatchSnapshot();
});

cases(
  'alter_table_add_column',
  async (opts) => {
    const [result] = await db.any(
      `
        select deparser.deparse( 
          ast_helpers.alter_table_add_column(
            v_schema_name := 'v_schema_name',
            v_table_name := 'v_table_name',
            v_column_name := 'v_column_name',
            v_column_type := $1::text
          )
        );
          `,
      [opts.name]
    );
    expect(result).toMatchSnapshot();
  },
  [
    { name: 'serial' },
    { name: 'bigserial' },
    { name: 'smallserial' },
    { name: 'bytea' },
    { name: 'smallint' },
    { name: 'int' },
    { name: 'bigint' },
    { name: 'char(1)' },
    { name: 'varchar(10)' },
    { name: 'text' },
    { name: 'boolean' },
    { name: 'inet' },
    { name: 'cidr' },
    { name: 'macaddr' },
    { name: 'numeric(1,0)' },
    { name: 'real' },
    { name: 'float(1)' },
    { name: 'float8' },
    { name: 'money' },
    { name: 'tsquery' },
    { name: 'tsvector' },
    { name: 'date' },
    { name: 'time' },
    { name: 'time with time zone' },
    { name: 'timestamp' },
    { name: 'timestamp with time zone' },
    { name: 'interval' },
    { name: 'bit' },
    { name: 'bit(4)' },
    { name: 'varbit' },
    { name: 'varbit(4)' },
    { name: 'box' },
    { name: 'circle' },
    { name: 'lseg' },
    { name: 'path' },
    { name: 'point' },
    { name: 'polygon' },
    { name: 'json' },
    { name: 'xml' },
    { name: 'uuid' },
    //
    { name: 'uuid[]' },
    { name: 'citext' },
    { name: 'geometry(point, 4326)' },
    { name: 'geometry(polygon, 4326)' }
  ]
);

cases(
  'types test',
  async (opts) => {
    const [result] = await db.any(
      `
      select ARRAY_LENGTH(REGEXP_MATCHES($1::text, '\\W'), 1) > 0 as contains;          `,
      [opts.name]
    );
    expect(result).toMatchSnapshot();
  },
  [
    { name: 'serial' },
    { name: 'bigserial' },
    { name: 'smallserial' },
    { name: 'bytea' },
    { name: 'smallint' },
    { name: 'int' },
    { name: 'bigint' },
    { name: 'char(1)' },
    { name: 'varchar(10)' },
    { name: 'text' },
    { name: 'boolean' },
    { name: 'inet' },
    { name: 'cidr' },
    { name: 'macaddr' },
    { name: 'numeric(1,0)' },
    { name: 'real' },
    { name: 'float(1)' },
    { name: 'float8' },
    { name: 'money' },
    { name: 'tsquery' },
    { name: 'tsvector' },
    { name: 'date' },
    { name: 'time' },
    { name: 'time with time zone' },
    { name: 'timestamp' },
    { name: 'timestamp with time zone' },
    { name: 'interval' },
    { name: 'bit' },
    { name: 'bit(4)' },
    { name: 'varbit' },
    { name: 'varbit(4)' },
    { name: 'box' },
    { name: 'circle' },
    { name: 'lseg' },
    { name: 'path' },
    { name: 'point' },
    { name: 'polygon' },
    { name: 'json' },
    { name: 'xml' },
    { name: 'uuid' },
    //
    { name: 'uuid[]' },
    { name: 'citext' },
    { name: 'geometry(point, 4326)' },
    { name: 'geometry(polygon, 4326)' }
  ]
);
