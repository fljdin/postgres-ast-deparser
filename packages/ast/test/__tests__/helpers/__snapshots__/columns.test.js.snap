// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`alter_table_add_column bigint 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name pg_catalog.bigint;",
}
`;

exports[`alter_table_add_column bigserial 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name pg_catalog.bigserial;",
}
`;

exports[`alter_table_add_column bit 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name bit;",
}
`;

exports[`alter_table_add_column bit(4) 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name pg_catalog.bit(4);",
}
`;

exports[`alter_table_add_column boolean 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name pg_catalog.boolean;",
}
`;

exports[`alter_table_add_column box 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name pg_catalog.box;",
}
`;

exports[`alter_table_add_column bytea 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name pg_catalog.bytea;",
}
`;

exports[`alter_table_add_column char(1) 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name pg_catalog.char(1);",
}
`;

exports[`alter_table_add_column cidr 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name pg_catalog.cidr;",
}
`;

exports[`alter_table_add_column circle 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name pg_catalog.circle;",
}
`;

exports[`alter_table_add_column citext 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name pg_catalog.citext;",
}
`;

exports[`alter_table_add_column date 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name pg_catalog.date;",
}
`;

exports[`alter_table_add_column float(1) 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name pg_catalog.float(1);",
}
`;

exports[`alter_table_add_column float8 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name pg_catalog.float8;",
}
`;

exports[`alter_table_add_column geometry(point, 4326) 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name pg_catalog.geometry(point, 4326);",
}
`;

exports[`alter_table_add_column geometry(polygon, 4326) 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name pg_catalog.geometry(polygon, 4326);",
}
`;

exports[`alter_table_add_column inet 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name pg_catalog.inet;",
}
`;

exports[`alter_table_add_column int 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name int;",
}
`;

exports[`alter_table_add_column interval 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name interval;",
}
`;

exports[`alter_table_add_column json 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name pg_catalog.json;",
}
`;

exports[`alter_table_add_column lseg 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name pg_catalog.lseg;",
}
`;

exports[`alter_table_add_column macaddr 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name pg_catalog.macaddr;",
}
`;

exports[`alter_table_add_column money 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name pg_catalog.money;",
}
`;

exports[`alter_table_add_column numeric(1,0) 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name pg_catalog.numeric(1,0);",
}
`;

exports[`alter_table_add_column path 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name pg_catalog.path;",
}
`;

exports[`alter_table_add_column point 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name pg_catalog.point;",
}
`;

exports[`alter_table_add_column polygon 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name pg_catalog.polygon;",
}
`;

exports[`alter_table_add_column real 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name pg_catalog.real;",
}
`;

exports[`alter_table_add_column serial 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name pg_catalog.serial;",
}
`;

exports[`alter_table_add_column smallint 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name pg_catalog.smallint;",
}
`;

exports[`alter_table_add_column smallserial 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name pg_catalog.smallserial;",
}
`;

exports[`alter_table_add_column text 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name pg_catalog.text;",
}
`;

exports[`alter_table_add_column time 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name time;",
}
`;

exports[`alter_table_add_column time with time zone 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name pg_catalog.time with time zone;",
}
`;

exports[`alter_table_add_column timestamp 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name timestamp;",
}
`;

exports[`alter_table_add_column timestamp with time zone 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name pg_catalog.timestamp with time zone;",
}
`;

exports[`alter_table_add_column tsquery 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name pg_catalog.tsquery;",
}
`;

exports[`alter_table_add_column tsvector 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name pg_catalog.tsvector;",
}
`;

exports[`alter_table_add_column uuid 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name pg_catalog.uuid;",
}
`;

exports[`alter_table_add_column uuid[] 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name pg_catalog.uuid[];",
}
`;

exports[`alter_table_add_column varbit 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name pg_catalog.varbit;",
}
`;

exports[`alter_table_add_column varbit(4) 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name pg_catalog.varbit(4);",
}
`;

exports[`alter_table_add_column varchar(10) 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name pg_catalog.varchar(10);",
}
`;

exports[`alter_table_add_column xml 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name pg_catalog.xml;",
}
`;

exports[`arrays 1`] = `
Object {
  "deparse": "ALTER TABLE v_schema_name.v_table_name ADD COLUMN v_column_name uuid[];",
}
`;

exports[`geometry 1`] = `
Object {
  "deparse": "geometry(polygon, 4326)",
}
`;

exports[`types test bigint 1`] = `undefined`;

exports[`types test bigserial 1`] = `undefined`;

exports[`types test bit 1`] = `undefined`;

exports[`types test bit(4) 1`] = `
Object {
  "contains": true,
}
`;

exports[`types test boolean 1`] = `undefined`;

exports[`types test box 1`] = `undefined`;

exports[`types test bytea 1`] = `undefined`;

exports[`types test char(1) 1`] = `
Object {
  "contains": true,
}
`;

exports[`types test cidr 1`] = `undefined`;

exports[`types test circle 1`] = `undefined`;

exports[`types test citext 1`] = `undefined`;

exports[`types test date 1`] = `undefined`;

exports[`types test float(1) 1`] = `
Object {
  "contains": true,
}
`;

exports[`types test float8 1`] = `undefined`;

exports[`types test geometry(point, 4326) 1`] = `
Object {
  "contains": true,
}
`;

exports[`types test geometry(polygon, 4326) 1`] = `
Object {
  "contains": true,
}
`;

exports[`types test inet 1`] = `undefined`;

exports[`types test int 1`] = `undefined`;

exports[`types test interval 1`] = `undefined`;

exports[`types test json 1`] = `undefined`;

exports[`types test lseg 1`] = `undefined`;

exports[`types test macaddr 1`] = `undefined`;

exports[`types test money 1`] = `undefined`;

exports[`types test numeric(1,0) 1`] = `
Object {
  "contains": true,
}
`;

exports[`types test path 1`] = `undefined`;

exports[`types test point 1`] = `undefined`;

exports[`types test polygon 1`] = `undefined`;

exports[`types test real 1`] = `undefined`;

exports[`types test serial 1`] = `undefined`;

exports[`types test smallint 1`] = `undefined`;

exports[`types test smallserial 1`] = `undefined`;

exports[`types test text 1`] = `undefined`;

exports[`types test time 1`] = `undefined`;

exports[`types test time with time zone 1`] = `
Object {
  "contains": true,
}
`;

exports[`types test timestamp 1`] = `undefined`;

exports[`types test timestamp with time zone 1`] = `
Object {
  "contains": true,
}
`;

exports[`types test tsquery 1`] = `undefined`;

exports[`types test tsvector 1`] = `undefined`;

exports[`types test uuid 1`] = `undefined`;

exports[`types test uuid[] 1`] = `
Object {
  "contains": true,
}
`;

exports[`types test varbit 1`] = `undefined`;

exports[`types test varbit(4) 1`] = `
Object {
  "contains": true,
}
`;

exports[`types test varchar(10) 1`] = `
Object {
  "contains": true,
}
`;

exports[`types test xml 1`] = `undefined`;
